src/lib/storage/dictionary_segment.cpp

```diff
@@ -0,0 +1,179 @@
+  auto values_count = values.size();
+
+  // Creates lookup.
+  std::unordered_map<T, ValueID> inverted_dictionary;
+  const auto dict_size = _dictionary.size();
+  for (auto dict_key = uint32_t{0}; dict_key < dict_size; ++dict_key) {
+    inverted_dictionary.insert({_dictionary[dict_key], ValueID{dict_key}});
+  }
+
+  DebugAssert(inverted_dictionary.size() == dict_size, "Dictionary contains duplicate elements.");
+
+  auto is_null = [&value_segment](ChunkOffset val_index) {
+    return value_segment->is_nullable() && value_segment->null_values()[val_index];
+  };
+
+  // Fills attribute vector with index of values in the dictionary.
+  for (auto val_index = ChunkOffset{0}; val_index < values_count; ++val_index) {
+    if (is_null(val_index)) {
+      _attribute_vector->set(val_index, null_value_id());
+    } else {
+      _attribute_vector->set(val_index, inverted_dictionary[values[val_index]]);
+    }
+  }
+}
+
+template <typename T>
+AllTypeVariant DictionarySegment<T>::operator[](const ChunkOffset chunk_offset) const {
+  return decompress(chunk_offset);
+}
+
+template <typename T>
+T DictionarySegment<T>::get(const ChunkOffset chunk_offset) const {
+  Assert(_attribute_vector->get(chunk_offset) != null_value_id(),
+         "Value at " + std::to_string(chunk_offset) + "is NULL.");
+  return decompress(chunk_offset);
+}
+
+template <typename T>
+std::optional<T> DictionarySegment<T>::get_typed_value(const ChunkOffset chunk_offset) const {
+  if (_attribute_vector->get(chunk_offset) == null_value_id()) {
+    return std::nullopt;
+  }
+  return decompress(chunk_offset);
+}
+
+template <typename T>
+T DictionarySegment<T>::decompress(const ChunkOffset chunk_offset) const {
+  return _dictionary[_attribute_vector->get(chunk_offset)];
+}
+
+template <typename T>
+const std::vector<T>& DictionarySegment<T>::dictionary() const {
+  return _dictionary;
+}
+
+template <typename T>
+std::shared_ptr<const AbstractAttributeVector> DictionarySegment<T>::attribute_vector() const {
+  return _attribute_vector;
+}
+
+template <typename T>
+ValueID DictionarySegment<T>::null_value_id() const {
+  return static_cast<ValueID>(_dictionary.size());
+}
+
+template <typename T>
+const T DictionarySegment<T>::value_of_value_id(const ValueID value_id) const {
+  DebugAssert(value_id != null_value_id(), "Value of value_id " + std::to_string(value_id) + " is null.");
+  return _dictionary[value_id];
+}
+
+template <typename T>
+ValueID DictionarySegment<T>::lower_bound(const T value) const {
+  auto lower_bound = std::lower_bound(_dictionary.begin(), _dictionary.end(), value);
+  if (lower_bound == _dictionary.end()) {
+    return INVALID_VALUE_ID;
+  }
+  return static_cast<ValueID>(lower_bound - _dictionary.begin());
+}
+
+template <typename T>
+ValueID DictionarySegment<T>::lower_bound(const AllTypeVariant& value) const {
+  if (variant_is_null(value)) {
+    return INVALID_VALUE_ID;
+  }
+  return lower_bound(type_cast<T>(value));
```

Comment:
```suggestion
  try {
    return lower_bound(type_cast<T>(value));
  } catch (...) {
    Fail("Tried lower_bound for inconvertible value to DictionarySegment.");
  }
```
If `value` holds in incompatible type to `T` an exception is thrown by `type_cast`. This should be caught and and handled with a Fail